kollect-core/src/main/kotlin/kollect/Cache.kt

```diff
@@ -1,45 +1,63 @@
 package kollect
 
+import arrow.Kind
 import arrow.core.Option
 import arrow.core.Tuple2
+import arrow.core.toOption
+import arrow.data.ListK
+import arrow.data.extensions.listk.foldable.foldable
+import arrow.data.k
+import arrow.typeclasses.Monad
+
+inline class DataSourceResult(val result: Any)
 
 /**
- * A `Cache` interface so the users of the library can provide their own cache.
+ *  Users of Kollect can provide their own cache by implementing this interface.
  */
-interface DataSourceCache {
-
-    fun <A> update(k: DataSourceIdentity, v: A): DataSourceCache
+interface DataSourceCache<F> {
 
-    fun <A> get(k: DataSourceIdentity): Option<A>
+    fun <I : Any, A : Any> lookup(i: I, ds: DataSource<I, A>): Kind<F, Option<A>>
```

Comment:
🤔 to me it has no sense to allow null ids (`I`) or null cached values (`A`), but pls let me know if you think I'm missing something.

Reply:
Let's see what @purrgrammer has to say about that :)

Reply:
I'm assuming you'd use `Option` to represent nullability, even though it is possible for `null` to be an `I` or an `A` at run time. If you can restrict the `Any` constraint more to avoid nullability I'd go ahead and do it :+1:

Reply:
@JorgeCastilloPrz the `Any` constraint isn't there in Fetch but lets leave this in.

@raulraja @pakoito do we have a policy for this in Arrow to restrict nullability in generics?

Reply:
I'm assuming you'd use `Option[I]` for allowing `None` values in `I`, even though `null` is a valid value for `I` at run time. If you can make a bigger constraint than `Any` to avoid `null` at run time I'd go for it :+1:

Reply:
Actually that makes a lot of sense now. But I'm not sure if we'd want to "pollute" the whole API contract with Option values just because they wanna use nullable indexes or value types.

Reply:
There's some implicit problem when I do that. I've been trying and it has some implications.

Any implementation of the`DataSourceCache` contract needs to type their value as a type `A` on their state (that means they require `A` in the class declaration). One example is `InMemoryCache`. 

When we make `I` and `A` not inherit from `Any`, we'd need to move at least `A` to the class level (`DataSourceCache<F, A>`) which could actually make sense. But then I'd personally also pull `I` up, so you have all the constraints in the class type `DataSourceCache<F, I, A>`.

That might sound fine, but that enforces to pass `I` and `A` all the way from the call sites, and you end up having client code on call sites like:

```kotlin
val io = Kollect.runEnv<ForIO, Int, Tuple2<ListK<Int>, ListK<Int>>>(
    IO.concurrent(), 
    IO.timer(EmptyCoroutineContext), 
    kollect(IO.concurrent()))

val res = io.fix().unsafeRunSync()
```

That's because the call site is not able to infer `I` by itself. Any ideas?

Reply:
`DataSourceCache[F, I, A]` wouldn't make sense since it would allow us only to store one type of identity-value pairs. `I` and `A` are parameters to the `insert` and `lookup` functions so the cache can be heterogeneous. See [here](https://github.com/47deg/fetch/blob/master/examples/src/test/scala/JedisExample.scala#L154-L185) for a Redis-backed implementation of `DataSourceCache` that can store heterogenous data types.

Regarding Fetch using [`DataSourceId` and other classes](https://github.com/47deg/fetch/blob/06556e7904c0d0533891a40f5f026b0220bfe5c5/shared/src/main/scala/cache.scala#L26) for wrapping cached values: I used it because someone suggested using [value classes](https://docs.scala-lang.org/overviews/core/value-classes.html) for wrapping data source identities and values, but note that those types don't exist at run time.

Reply:
We can do the same with inline classes, which I actually removed for simplicity, but that would not save us from the `Any` stuff. If you hadn't those classes in Scala you'd have typed your state on the `InMemoryCache` as  `Map[(String, Any), Any]` (given current Fetch constraints on those types)

Reply:
>If you hadn't those classes in Scala you'd have typed your state on the InMemoryCache as Map[(String, Any), Any] (given current Fetch constraints on those types)

Exactly

Reply:
If we need Option to push this, let's use it for the time being and revisit this when we have a more stable API

Reply:
how would `Option<A>` solve the issue?, you'd still need to provide a type for `A` in the `InMemoryCache `, and that's gonna be `Any`, since we don't have `A` in the type declaration, but until cache methods are called.

Reply:
Added #15 so we can keep the discussion there about how to iterate on the issue.